obvs/patchscope.py

```diff
@@ -227,20 +248,69 @@ def target_forward_pass(self) -> None:
+            self.target.layer
+        ]
+
+        # if a head index is given, need to access the ATTN and HEAD components
+        if self.target.head is not None:
+            attn = getattr(layer, self.ATTN_TARGET)
+            # TODO may not be .input for other models
+            concat_head_act = getattr(attn, self.HEAD_TARGET).input[0][0]
+
+            # need to reshape the output of head into the specific heads
+            split_head_act = einops.rearrange(
+                concat_head_act, 'batch pos (n_head d_head) -> batch pos n_head d_head',
+                n_head=attn.num_heads, d_head=attn.head_dim
+            )
+
+            # check if the dimensions of the mapped_hidden_state and target head activations match
+            target_act = split_head_act[:, self._target_position, self.target.head, :]
+            if self._mapped_hidden_state.shape != target_act.shape:
+                logger.error('Cannot set activation of head %s in target model with shape'
+                             ' %s to patched activation of source model with shape %s!',
+                             self.target.head, list(target_act.shape),
+                             list(self._mapped_hidden_state.shape))
+                return
+            split_head_act[:, self._target_position, self.target.head, :] = self._mapped_hidden_state
+        else:
+            layer.output[0][:, self._target_position, :] = self._mapped_hidden_state
 
         self._target_outputs.append(self.target_model.lm_head.output[0].save())
         for _ in range(self.target.max_new_tokens - 1):
             self._target_outputs.append(self.target_model.lm_head.next().output[0].save())
 
+    def check_patchscope_setup(self) -> bool:
+        """ Check if patchscope is correctly set-up before running """
+
+        # head can be int or None, patchscope run is only possible if they have the same type
+        # TODO: Find out how to do it PEP8 compliant
+        if type(self.source.head) != type(self.target.head):
+            logger.error('Cannot run patchscope with source head attribute: %s and target'
+                         ' head attribute: %s. Both need to be of the same type.', self.source.head,
+                         self.target.head)
+            return False
+
+        # currently, accessing single head activations is only supported for GPT2LMHead models
+        if (self.source.head is not None and 'gpt2' not in self.source.model_name or
+                self.target.head is not None and 'gpt2' not in self.target.model_name):
+            raise NotImplementedError('Accessing single head activations is currently only'
+                                      ' implemented for GPT2-style models')
+
+        return True
+
     def run(self) -> None:
         """
         Run the patchscope
         """
+
+        # check before running
+        if not self.check_patchscope_setup():
+            logger.error('Abort running patchscope')
+            return
```

Comment:
Should this be an exception?

It's better to alert users early than for them to to receive garbage results.

Reply:
Same here, good idea to raise an exception